pontoon/pretranslation/pretranslate.py

```diff
@@ -1,130 +1,257 @@
+                    keys[idx] == "one" for idx in plural_selectors
+                ):
+                    rm_keys.append(keys)
+                else:
+                    pattern[:] = self.pattern(pattern)
+            for keys in rm_keys:
+                del msg.variants[keys]
+
+            # Copy catchall patterns for the locale's other plural categories
+            if plurals:
+                for idx in plural_selectors:
+                    tgt_variants = {}
+                    for keys, pattern in msg.variants.items():
+                        key = keys[idx]
+                        if isinstance(key, CatchallKey):
+                            for pc in plurals:
+                                pc_keys = keys[:idx] + (pc,) + keys[idx + 1 :]
+                                if pc_keys not in msg.variants:
+                                    tgt_variants[pc_keys] = deepcopy(pattern)
+                        tgt_variants[keys] = pattern
+                    msg.variants = tgt_variants
+
+    def pattern(self, pattern: Pattern) -> Pattern:
+        # First try to get a 100% match from Translation Memory
+        tm_source = (
+            "".join(
+                el.value
+                if isinstance(el, FTL.TextElement)
+                else serialize_expression(el)
+                for el in fluent_astify_message(PatternMessage(pattern)).elements
+            )
+            if self.format == Format.fluent
+            else self.source
         )
-        if pretranslation is None:
-            return None
-        return (pretranslation, services[service])
-
-
-def get_pretranslated_data(source: str, locale: Locale, preserve_placeables: bool):
-    # Empty strings and strings containing whitespace only do not need translation
-    if re.search("^\\s*$", source):
-        return source, "tm"
-
-    # Try to get matches from Translation Memory
-    tm_response = get_translation_memory_data(text=source, locale=locale)
-    tm_perfect = [t for t in tm_response if int(t["quality"]) == 100]
-    if tm_perfect:
-        return tm_perfect[0]["target"], "tm"
-
-    # Fetch from Google Translate
-    elif locale.google_translate_code:
-        gt_response = get_google_translate_data(
-            text=source, locale=locale, preserve_placeables=preserve_placeables
+        if not tm_source or tm_source.isspace():
+            return pattern
+        tm_q100 = list(
+            TranslationMemoryEntry.objects.filter(
+                locale=self.locale, source=tm_source
+            ).values_list("target", flat=True)
         )
-        if gt_response["status"]:
-            return gt_response["translation"], "gt"
+        if tm_q100:
+            tm_best = max(set(tm_q100), key=tm_q100.count)
+            self.services.append("tm")
+            if self.format == Format.fluent:
+                te = fluent_parse_entry(f"key = {tm_best}\n")
+                assert isinstance(te.value, PatternMessage)
+                return te.value.pattern
```

Comment:
This feels a bit hacky as well, and probably should be supported by moz.l10n.

Reply:
Should file a bug, too.